### ISSUE TITLE
Support for Embarcaderos bcc32c Clang based Compiler

### ISSUE THREAD

**Issue by Raensi** [fastbuild/fastbuild#762]
Hello, first i like to thank you, for the great work you guys do.

Now to my problem.
I tried to integrate the **bcc32c / bcc32x Compiler from Embarcadero**. But it seems, that Fastbuild has some problems with working with that compiler.
First to get it to work, i had to disable the check for the "-c" Flag in the FunctionObjectList.cpp File, because the bcc32c Compiler works with the -cc1 Flag. Then i got it to work.
Sadly it is only working with the **custom** Compiler option. So there is no support for caching and distribution.
Is it possible to add support for this compiler too?

**Comment by ffulin** [comment]
I'm not familiar with this compiler, but I spent a bit of time looking for some documentation. I found something (http://docs.embarcadero.com/products/rad_studio/delphiAndcpp2009/HelpUpdate2/EN/html/devwin32/bcc32_xml.html), so my thoughts are based on that. Let me know if you have a better reference.

From those docs, it seems:

**-c option**

-c is a supported compiler option. If it's not desirable to enforce -c being present, then certainly FASTBuild can be modified as needed. (We can skip that check for bcc32* for example, or check for other alternative args)
 
**custom compiler**

FASTBuild detects compilers by name and alters behavior as needed. For compilers it doesn't understand, then there is not distributed compilation or caching as you noted. Additionally, dependencies are not correctly tracked, so things will not rebuild correctly. This means "custom" is usually only useful for one off tools that have simple one file in/one file out relationships.

**adding basic support**

FASTBuild can support any compiler that can emit dependency information if give some command line args. It most commonly does this by using the preprocessor (the -E command line). According to the docs I found bcc32* doesn't seem to support that. The only thing I found was a reference to the -mo option. If that can output dependency information, FASTBuild could be modified to support the compiler. As no other compiler works this way currently, it will be a bit more work than adding a compiler that follows gcc/clang or msvc conventions.

**adding distributed compilation support**

Distributed compilation requires the compiler be able to be run in a preprocessor only mode (-E as noted above). I couldn't find any reference to an option like this. Without this option or an equivalent, it's not currently possible to support distributed compilation.

**adding caching support**

FASTBuild usually uses the preprocessor for caching as well (-E as noted above). However, it is possible to the the new experimental LightCache feature for caching which is compiler independent, so that may be an option.

**Comment by ffulin** [comment]
If the compiler is truly clang based, then it might be possible that with the correct command line options that we could set CompilerFamily to "clang" and have everything work.

**Comment by Raensi** [comment]
Yeah, the -c option is supportet. But we have to run it with the -cc1 option instead of the -c one. This calls the frontend of the compiler and not the driver.

Here you can see the difference between the compilers:
[http://docwiki.embarcadero.com/RADStudio/Rio/en/Clang-enhanced_C%2B%2B_Compilers](http://docwiki.embarcadero.com/RADStudio/Rio/en/Clang-enhanced_C%2B%2B_Compilers)

We are using the bcc32**c** Compiler.
You can find more information here
[http://docwiki.embarcadero.com/RADStudio/Rio/en/Win32_Clang-enhanced_Compilers](http://docwiki.embarcadero.com/RADStudio/Rio/en/Win32_Clang-enhanced_Compilers)

I tried with the clang flag. But apparently the compiling with the -E flag does not work as it should be. Because it creates a file with no extention with a given size, but the object file wont be built. Therefore the Assertion in the ObjectNode.cpp (782) is thrown. output && outputsize
![File](https://user-images.githubusercontent.com/26414864/86272829-814aaa80-bbcf-11ea-960a-97bbbea57ef5.png)

**Comment by ffulin** [comment]
According to the first link you provided:

> Clang-enhanced C++ compilers are also preprocessors when run with the -E switch. 

With that in mind, it seems it should be possible to detect use of the compiler and use that switch. It would need some investigation to figure out why that doesn't simply work when setting the compiler family.
